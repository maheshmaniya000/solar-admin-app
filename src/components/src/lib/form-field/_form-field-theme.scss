@use '~@angular/material' as mat;

@mixin theme($theme-or-color-config) {
  $color-config: mat.get-color-config($theme-or-color-config);
  $primary: map-get($color-config, primary);
  $warn: map-get($color-config, warn);
  $foreground: map-get($color-config, foreground);
  $is-dark-theme: map-get($color-config, is-dark);
  $underline-color-base: mat.get-color-from-palette($foreground, divider, if($is-dark-theme, 1, 0.87));
  $underline-focused-color: mat.get-color-from-palette($primary, text);

  mat-form-field.mat-form-field-appearance-outline {
    .mat-form-field-subscript-wrapper .mat-hint {
      color: mat.get-color-from-palette($foreground, hint-text);
    }

    &,
    &.mat-form-field-disabled {
      .mat-form-field-outline {
        color: mat.get-color-from-palette($foreground, disabled);
      }
    }

    &,
    &.mat-focused {
      .mat-form-field-outline-thick {
        color: mat.get-color-from-palette($foreground, icon);
      }
    }

    &.mat-form-field-invalid {
      &[doNotAppendFormError] {
        .mat-form-field-outline-thick,
        &.mat-focused .mat-form-field-outline-thick {
          color: $underline-focused-color;
        }
      }

      ::placeholder {
        color: mat.get-color-from-palette($warn);
      }
    }
  }
}
