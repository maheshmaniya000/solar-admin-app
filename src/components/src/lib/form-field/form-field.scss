@use '../icon/icon' as icon;

$horizontal-spacing: 8px;

mat-form-field {
  width: 100%;

  &.mat-form-field {
    display: block;
  }

  &.mat-form-field-appearance-outline {
    ::placeholder {
      transition: none;
      font-style: italic;
    }

    .mat-form-field-wrapper {
      margin: 0;
      padding-bottom: 0;
    }

    .mat-form-field-outline {
      top: 0;
    }

    .mat-form-field-outline-start {
      border-radius: 4px 0 0 4px;
    }

    .mat-form-field-outline-end {
      border-radius: 0 4px 4px 0;
    }

    .mat-form-field-outline-thick {
      .mat-form-field-outline-start,
      .mat-form-field-outline-end,
      .mat-form-field-outline-gap {
        border-width: 1px;
      }
    }

    &[doNotAppendFormError].mat-form-field-invalid {
      &:not(.mat-focused) .mat-form-field-outline {
        opacity: 1;

        &.mat-form-field-outline-thick {
          opacity: 0;
        }
      }
    }

    .mat-form-field-flex {
      align-items: center;
      margin-top: 0;
      padding: 0 $horizontal-spacing;
    }

    .mat-form-field-infix {
      padding: 5px 0;
      border-top: none;
    }

    .mat-form-field-subscript-wrapper {
      position: static;
      top: 0;
      margin-top: 4px;
      padding: 0;
      font-size: 100%;
    }

    &.ng-untouched,
    &.ng-valid,
    &.ng-pending {
      &.no-hint .mat-form-field-subscript-wrapper {
        display: none;
      }
    }

    .mat-form-field-prefix,
    .mat-form-field-suffix {
      position: static;

      .mat-icon {
        @include icon.size(default);
      }
    }

    &:not(.mat-form-field-appearance-legacy) {
      .mat-form-field-prefix,
      .mat-form-field-suffix {
        .mat-icon-button {
          display: block;
          width: auto;
          height: auto;
        }
      }
    }

    .mat-form-field-prefix {
      margin-right: $horizontal-spacing;
    }

    .mat-form-field-suffix {
      margin-left: $horizontal-spacing;
    }

    .mat-form-field-label-wrapper {
      display: none;
    }
  }
}
