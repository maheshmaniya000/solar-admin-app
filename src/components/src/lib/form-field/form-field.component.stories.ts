import { APP_BASE_HREF, CommonModule } from '@angular/common';
import { HttpClientModule } from '@angular/common/http';
import { FormsModule } from '@angular/forms';
import { MatButtonModule } from '@angular/material/button';
import { MatIconModule } from '@angular/material/icon';
import { MatInputModule } from '@angular/material/input';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import { moduleMetadata, Story } from '@storybook/angular';

import { TranslocoRootModule } from 'ng-shared/core/transloco-root.module';

import { SgDefaultsModule } from '../defaults/defaults.module';
import { FormErrorModule } from '../form-error/form-error.module';
import { FormFieldModule } from './form-field.module';

export default {
  title: 'Solargis/Components/Form Field',
  decorators: [
    moduleMetadata({
      imports: [
        BrowserAnimationsModule,
        MatInputModule,
        SgDefaultsModule,
        FormFieldModule,
        FormsModule,
        CommonModule,
        MatIconModule,
        MatButtonModule,
        HttpClientModule,
        TranslocoRootModule,
        FormErrorModule
      ],
      providers: [{ provide: APP_BASE_HREF, useValue: '.' }]
    })
  ]
};

export const bottomSpacing: Story = props => ({
  props,
  template: `
    <div class="container">
      <label for="input1">Hint</label>
      <div>
        <mat-form-field>
          <input id="input1" matInput placeholder="placeholder">
          <mat-hint *ngIf="hintVisible">Some hint</mat-hint>
        </mat-form-field>
        <div>Some block to demonstrate form field height without hint.</div>
      </div>

      <label for="input2">Error</label>
      <div>
        <mat-form-field>
          <input id="input2" matInput placeholder="placeholder" ngModel required>
        </mat-form-field>
        <div>
          Some block to demonstrate form field height with dynamic error. Since
          this field is required, you can see its error when its value is empty,
          and you interact with the input at least once (click in and out).
        </div>
      </div>

      <label for="input3">Error + hint</label>
      <div>
        <mat-form-field>
          <input id="input3" matInput placeholder="placeholder" ngModel required>
          <mat-hint *ngIf="hintVisible">Some hint</mat-hint>
        </mat-form-field>
        <div>
          Some block to demonstrate form field height with dynamic error & hint. Since
          this field is required, you can see its error when its value is empty,
          and you interact with the input at least once (click in and out).
        </div>
      </div>
    </div>
  `,
  styles: [
    `
    .container {
      display: grid;
      grid-template-columns: repeat(2, 1fr);
      grid-gap: 10px;
      width: 800px;
    }
  `
  ]
});

bottomSpacing.argTypes = {
  hintVisible: {
    name: 'Hint visible',
    control: { type: 'boolean' },
    defaultValue: false
  }
};

export const disabled: Story = props => ({
  props,
  template: `
    <div class="container">
      <p>
        'Required' error message will be rendered only if this input is touched
         and enabled. To demonstrate that disabled inputs do not display error, first
         interact with it (click in and out) and then set the disabled control to
         'True'.
      </p>
      <mat-form-field>
        <input matInput ngModel [disabled]="disabled" required placeholder="placeholder"/>
      </mat-form-field>
    </div>
  `,
  styles: [
    `
    .container {
      width: 500px;
    }
  `
  ]
});

disabled.argTypes = {
  disabled: {
    name: 'Disabled',
    control: { type: 'boolean' },
    defaultValue: false
  }
};

export const prefixAndSuffix: Story = () => ({
  template: `
    <div class="container">
      <label for="input1">Text</label>
      <mat-form-field>
        <input id="input1" matInput placeholder="placeholder"/>
        <span matPrefix>+421</span>
        <span matSuffix>kWh</span>
      </mat-form-field>
      <label for="input2">Icon / Icon button</label>
      <mat-form-field>
        <input id="input2" matInput placeholder="placeholder"/>
        <mat-icon matPrefix>search</mat-icon>
        <button mat-icon-button color="primary" matSuffix>
          <mat-icon>visibility_off</mat-icon>
        </button>
      </mat-form-field>
    </div>
  `,
  styles: [
    `
    .container {
      display: grid;
      grid-template-columns: repeat(2, 1fr);
      grid-gap: 10px;
      width: 500px;
    }
  `
  ]
});
