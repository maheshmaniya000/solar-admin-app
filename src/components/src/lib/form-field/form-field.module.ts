import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { MatFormFieldModule } from '@angular/material/form-field';
import { TranslocoModule } from '@ngneat/transloco';

import { NoHintDirective } from './no-hint/no-hint.directive';

@NgModule({
  declarations: [NoHintDirective],
  imports: [CommonModule, MatFormFieldModule, TranslocoModule],
  exports: [MatFormFieldModule, NoHintDirective]
})
export class FormFieldModule {}
