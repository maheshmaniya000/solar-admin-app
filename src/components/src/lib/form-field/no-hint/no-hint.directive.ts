import { ContentChild, Directive, HostBinding } from '@angular/core';
import { MatHint } from '@angular/material/form-field';
import { isNil } from 'lodash-es';

@Directive({
  // eslint-disable-next-line @angular-eslint/directive-selector
  selector: 'mat-form-field'
})
export class NoHintDirective {
  @ContentChild(MatHint) hint: MatHint;
  @HostBinding('class.no-hint')
  get noHintClassApplied(): boolean {
    return isNil(this.hint);
  }
}
