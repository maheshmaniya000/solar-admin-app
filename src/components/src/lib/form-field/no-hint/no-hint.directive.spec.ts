import { MatHint } from '@angular/material/form-field';
import { createDirectiveFactory, SpectatorDirective } from '@ngneat/spectator/jest';
import { isNil } from 'lodash-es';
import { MockComponents } from 'ng-mocks';

import { NoHintDirective } from './no-hint.directive';

describe('NoHintDirective', () => {
  describe('unit', () => {
    let directive: NoHintDirective;

    beforeEach(() => {
      directive = new NoHintDirective();
    });

    describe('noHintClassApplied', () => {
      const mockHint = { id: 'some-hint-id', align: 'start' };
      (
        [
          {
            hint: mockHint,
            expectedResult: false
          },
          {
            hint: undefined,
            expectedResult: true
          }
        ] as {
          hint: MatHint;
          expectedResult: boolean;
        }[]
      ).forEach(({ hint, expectedResult }) =>
        it(`should return ${expectedResult} when hint is ${isNil(hint) ? 'undefined' : 'defined'}`, () => {
          directive.hint = hint;
          expect(directive.noHintClassApplied).toBe(expectedResult);
        })
      );
    });

    describe('component', () => {
      const createDirective = createDirectiveFactory({
        directive: NoHintDirective,
        declarations: [MockComponents(MatHint)],
        shallow: true
      });
      let spectator: SpectatorDirective<NoHintDirective>;

      it(`should match snapshot with no-hint class that indicates the form
 field does not contain hint when it really does not contain any mat-hint element`, () => {
        spectator = createDirective(`
            <mat-form-field></mat-form-field>
          `);
        expect(spectator.fixture).toMatchSnapshot();
      });

      it(`should match snapshot without no-hint class when the form really
 contains a mat-hint element`, () => {
        spectator = createDirective(`
            <mat-form-field>
              <mat-hint></mat-hint>
            </mat-form-field>
          `);
        expect(spectator.fixture).toMatchSnapshot();
      });
    });
  });
});
