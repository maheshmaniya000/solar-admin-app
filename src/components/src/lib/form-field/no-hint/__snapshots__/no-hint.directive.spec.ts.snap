// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoHintDirective unit component should match snapshot with no-hint class that indicates the form
 field does not contain hint when it really does not contain any mat-hint element 1`] = `
<ng-component
  __ngContext__={[Function LRootView]}
>
  <mat-form-field
    class="no-hint"
  />
</ng-component>
`;

exports[`NoHintDirective unit component should match snapshot without no-hint class when the form really
 contains a mat-hint element 1`] = `
<ng-component
  __ngContext__={[Function LRootView]}
>
  <mat-form-field>
    <mat-hint />
  </mat-form-field>
</ng-component>
`;
